pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -399,6 +400,8 @@ private static DataType getDataType(Expression expression, Schema schema) {
         dataType = DataType.INT;
       } else if ("VARCHAR".equals(targetTypeLiteral)) {
         dataType = DataType.STRING;
+      } else if (targetTypeLiteral.endsWith("_ARRAY")) {
```

Comment:
Good catch! Should we perform this before getting into the if checks? This way we can also handle things like `INTEGER_ARRAY` etc

Reply:
Those should still be handled right since the above if checks are equals checks? It shouldn't matter either way though so I've re-ordered the if statements.

Reply:
I think Jackie means to first check the `_ARRAY` suffix. If it is the case, modify `targetTypeLiteral` to do not include it. Then apply all other ifs (like `"INTEGER".equal(targetTypeLiteral)`). 

Then when the original `targetTypeLiteral` was `INTEGER_ARRAY` we would return `DataType.INT`. In the current state, `INTEGER_ARRAY` would fail because `INTEGER` is not a DataType literal.

Reply:
Ah, I see what you mean. However, this method is called on the server side in the v2 engine, after the query is already compiled I believe. In v1, this optimizer is called in the broker itself, before dispatch to servers. I'm not quite familiar yet with how the type system and conversion works in v2 with Calcite <-> Pinot types - but when this method is called, the type in the `cast` function is already converted to the "Pinot type". So for instance, if the query has `CAST(mv_col AS BIGINT ARRAY)`, it becomes `LONG_ARRAY`. If the query has `CAST(mv_col AS INTEGER ARRAY)`, it becomes `INT_ARRAY`. That's why the integration test I added with `CAST(DivAirportIDs AS BIGINT ARRAY)` works fine even though the `DataType` enum doesn't have a `BIGINT` value. I've tested manually with `CAST(mv_col AS INTEGER ARRAY)` as well and it works just fine. I definitely might be missing something here though since I don't have a clear picture of all the moving parts involved just yet.

Reply:
Hmm but before this fix, v1 didn't support `ARRAY` keyword in cast function at all - MV fields were cast the exact same way as SV fields. So `CAST(mv_col AS INT)` and `CAST(mv_col AS INTEGER)` both work whereas none of `CAST(mv_col AS INT ARRAY)` / `CAST(mv_col AS INTEGER ARRAY)` / `CAST(mv_col AS INT_ARRAY)` / `CAST(mv_col AS INTEGER_ARRAY)` work (all result in the error: `No enum constant org.apache.pinot.spi.data.FieldSpec.DataType.XYZ`). 

And in the v2 query engine, before this fix, there was no way to cast MV columns to a numeric type that works. After this fix, `CAST(mv_col AS INT ARRAY)` / `CAST(mv_col AS INTEGER ARRAY)` both work but `CAST(mv_col AS INT_ARRAY)` / `CAST(mv_col AS INTEGER_ARRAY)` don't work since `INT_ARRAY` / `INTEGER_ARRAY` aren't recognized by Calcite. Inadvertently, this fix also makes `CAST(mv_col AS INT_ARRAY)` work in the v1 query engine, but as you pointed out `CAST(mv_col as INTEGER_ARRAY)` won't work. Note that even `CAST(mv_col AS INT ARRAY)` won't work in the v1 query engine (with or without this fix).

This is fairly messy but I think the question boils down to whether we want to support `ARRAY` keyword for MV columns in the cast function in the v1 query engine and whether we're okay with introducing an additional difference between the two query engines (`INT_ARRAY` in v1 vs `INT ARRAY` in v2)?

Reply:
Ah never mind but the issue in v1 was due to this [NumericalFilterOptimizer](https://github.com/apache/pinot/blob/99f6934166633308547d37cacc634e03c723ab17/pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java#L385) itself so things like `CAST(mv_col AS STRING_ARRAY)` did work in v1 before this fix. I think it makes sense to support `INTEGER_ARRAY` as well as `INT_ARRAY` in v1 👍